global.json

```diff
@@ -0,0 +1,6 @@
+{
+    "sdk": {
+        "version": "3.1.401",
+        "rollForward": "latestPatch"
+    }
+}
```

Comment:
Looking at the [`global.json` documentation](https://docs.microsoft.com/en-us/dotnet/core/tools/global-json?tabs=netcore3x), it says:

> In general, you want to use the latest version of the SDK tools, so no *global.json* file is needed. In some advanced scenarios, you might want to control the version of the SDK tools, and this article explains how to do this.

Not sure if we fall into the "some advanced scenarios" part (I haven't looked at the failing test). If it's not too hard to fix the tests with .NET 5 in a way that continues to work with 3.1, would that be better?

Reply:
I think the failing test is a good example of the types of problems that might arise when the `global.json` is not pinned.

I believe what you want to have here, is a state of the repo that gives you deterministic building and testing conditions. Otherwise contributor A might build with `3.1.100`, contributor B might build with `3.1.401` and contributor C might build with `5.0.100-rc.1` and potentially get subtly (or not so subtly) different behaviors. There are different MSBuild versions in those SDKs (16.3 vs 16.7 vs 16.8) and from what I can say from the experience of building C# tooling they can have quite a lot of behavioral differences. Given that this project relies on a custom project SDK makes it even more important. 

We did the same thing in OmniSharp - where the SDK is pinned to the latest, guaranteeing the contributors are all on the same one https://github.com/OmniSharp/omnisharp-roslyn/blob/master/global.json 

It also gives you clear, unambiguous context: this is the only dotnet SDK version supported for contributing to the repo, and it will be changed to a different one only by a conscious version bump in `global.json`, so you are saving yourself and others time trying to figure out why something fails on a newer or older version.

Reply:
Yeah, it makes sense, I agree that reproducible builds are important.

It looks like when the pinned version is not installed, the build fails rather than automatically downloading the right version, is that correct? That is one difference between pinning the SDK and pinning NuGet/NPM package versions, I think, is needing to manually install the right version rather than having the build automatically download it. That's not a reason not to use global.json necessarily, but just something I thought about.

@anpaz @bettinaheim Do you have any other thoughts on using global.json?

Reply:
correct, the .NET SDK behavior is such that when the relevant version is missing, it fails with a message that the SDK is missing.

The behavior can be additionally controlled by `"rollForward"` setting:
 - setting it in `global.json` to `latestPatch` would allow any newer SDK that still fits 3.1.4xx to  be used too (but not e.g. 3.1.500)
 - setting it to `latestFeature` would allow any newer SDK that fits into `3.1.xxx` to be used as well (but not e.g. 5.0.0). 
 - setting it to `major` would allow any newer SDK to be used

It depends on the level of SDK predictability you want to have. For example in OmniSharp we chose to pin the exact version, while e.g. on Roslyn there is the lowest SDK defined and `rollForward` is set to `major`.